Refactor FXIOS-8018 [v123] Remove Deferred usage from InitialViewController

bmihai23 opened the pull request:
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-8018)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/17889)

## :bulb: Description
- Replaced the **`Deferred<ExtensionUtils.ExtractedShareItem?>`** return type in **`getShareItem`** with a completion handler.
- Refactored the extraction of shared items to use a callback for asynchronous handling.

## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed I updated documentation / comments for complex code and public methods

adudenamedruby reviewed (approved):
Looks great! Tyvm for your contribution!

isabelrios commented:
There is a UI test that is failing consistently in this PR: 
`testSharePageWithShareSheetOptions` https://github.com/mozilla-mobile/firefox-ios/blob/89327acb46496b278a84ad01000f765fbed661e7/firefox-ios/firefox-ios-tests/Tests/XCUITests/PhotonActionSheetTests.swift#L82

When clicking on Fennec here, the share sheet is not open, nothing happens.
<img width="216" alt="Screenshot 2023-12-27 at 15 57 56" src="https://github.com/mozilla-mobile/firefox-ios/assets/1897507/5d8b463a-57de-4707-b7c7-1dc3ffd53ccc">

Please make sure this test passes before landing this PR. Thanks!

bmihai23 commented:
> There is a UI test that is failing consistently in this PR: `testSharePageWithShareSheetOptions`
> 
> https://github.com/mozilla-mobile/firefox-ios/blob/89327acb46496b278a84ad01000f765fbed661e7/firefox-ios/firefox-ios-tests/Tests/XCUITests/PhotonActionSheetTests.swift#L82
> 
> When clicking on Fennec here, the share sheet is not open, nothing happens. <img alt="Screenshot 2023-12-27 at 15 57 56" width="216" src="https://private-user-images.githubusercontent.com/1897507/293053466-5d8b463a-57de-4707-b7c7-1dc3ffd53ccc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDM3MDQ0MDcsIm5iZiI6MTcwMzcwNDEwNywicGF0aCI6Ii8xODk3NTA3LzI5MzA1MzQ2Ni01ZDhiNDYzYS01N2RlLTQ3MDctYjdjNy0xZGMzZmZkNTNjY2MucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMjdUMTkwODI3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDU2NDFkOThhMzc2MzYyZDkzNGQzY2E5YWZhYTVjODIyY2Q2OGQxNzkyYjNjOGRkOTNhYWI2MzU1ODU0NTMyZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.GrNtJmSMSy3kDE3k-0UKcTPn4_payzwq8y3dhFpt5ZY">
> 
> Please make sure this test passes before landing this PR. Thanks!

Thank you for bringing this to my attention. I have addressed the reported issue related to the failing UI test (`testSharePageWithShareSheetOptions`). The problem was identified and fixed in the code, ensuring that clicking on "Fennec" now correctly opens the share sheet as expected. Thanks!
